### ISSUE TITLE
conan_get_compiler function does not work properly

### ISSUE THREAD

**Issue by OlegWorld** [conan-io/conan#6658]
the function conan_get_compiler, which is defined in conanbuildinfo.cmake works as follows:
- opens conaninfo.txt
- tries to make a regex match of compiler and compiler.version
but if i follow the instructions of this manual during package creation: https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html
especially this example:
```
def package_id(self):
        v = Version(str(self.settings.compiler.version))
        if self.settings.compiler == "gcc" and (v >= "4.5" and v < "5.0"):
            self.info.settings.compiler.version = "GCC version between 4.5 and 5.0"
```
then conan_get_compiler function sets CONAN_INFO_COMPILER_VERSION as "GCC", which leads to the following cmake error:
```
-- Conan: Checking correct version: 4.7
CMake Error at conanbuildinfo.cmake:526 (message):
  Detected a mismatch for the compiler version between your conan profile
  settings and CMake:

  Compiler version specified in your conan profile: GCC

  Compiler version detected in CMake: 4.7

  Please check your conan profile settings (conan profile show
  [default|your_profile_name])
```

**Comment by memsharded** [comment]
I am trying a fix here: https://github.com/conan-io/conan/pull/6659

**Comment by memsharded** [comment]
This has been fixed in https://github.com/conan-io/conan/pull/6659, already merged. It will be released next 1.24, thanks very much for reporting this issue!

**Comment by memsharded** [comment]
Tip: packages already merged in the develop branch can be tested from Test PyPI:

$ pip install --index-url https://test.pypi.org/simple/ conan


(once CI finish and uploads it)
